### PR TITLE
Give the bare `{children}` splice a nullish guard

### LINKED ISSUE #2775: An unguarded `{props.children}` renders the literal text `undefined` on CSR mount when the caller passes no children (SSR renders empty)
## The divergence

A child component that interpolates `props.children` without a guard, called by a parent that passes no children, renders **differently on the SSR leg and the CSR-mount leg**. SSR renders an empty body; CSR mount renders the four-character string `undefined` into the DOM.

Source — nothing exotic, this is the ordinary shape:

```tsx
export function Card(props: { header?: any; children?: any }) {
  return (
    <section>
      <header>{props.header}</header>
      <div class="body">{props.children}</div>
    </section>
  )
}

export function Parent() {
  return <Card header={<span>HEADER</span>} />
}
```

## Measured

**Client template** (`Card.client.js`, Hono compile):

```js
hydrate('Card', { init: initCard, template: (_p) => `…<header bf="s1">…${escapeTextOrMarkup(_p.header)}…</header><div class="body">${_p.children}</div></section>` })
```

`children` reaches the template literal as a bare splice. Evaluating that lambda the way the mount path does, with no `children` in `_p`:

```
unguarded, no children : "<div class=\"body\">undefined</div>"
guarded,   no children : "<div class=\"body\"></div>"
```

**SSR, same source** — both render nothing:

```jsx
// Hono (reference)
<div className="body">{props.children}</div>
```
```twig
{# Twig #}
<div class="body">{{ bf.string(children) }}</div>
```

Hono's JSX runtime renders `undefined` as nothing; the DSL adapters route the value through `bf.string(...)`. Only the client template stringifies it raw.

So the three-way contract is broken on the SSR ≡ CSR-mount leg: SSR-only and SSR+hydration render an empty body, pure CSR mount renders `undefined`.

## The asymmetry that makes it visible

In the same emitted template, the sibling prop is not spliced raw:

| prop | emitted as |
|---|---|
| `header` | `${escapeTextOrMarkup(_p.header)}` |
| `children` | `${_p.children}` |

`escapeTextOrMarkup(undefined)` returns `""`, so `header` is already safe for the same input. Only the reserved children slot takes the bare path.

Whether `children` should also be *escaped* is a separate question and this issue does not claim it should — a JSX children payload is markup, and escaping it would be wrong. The defect is the missing **undefined** handling, not the missing escape.

## How it was found

While reviewing #2774 (the #2773 children-slot predicate fix). That PR's new fixture `jsx-element-prop-no-children` writes its `Card` as `{props.children ?? ''}`, and its docstring records why: the unguarded form trips this quirk, which is not what that fixture exists to pin. The pre-existing `component-with-jsx-children` fixture uses the same guard.

That convention is reasonable per-fixture, but it means **the corpus has been routing around this defect rather than pinning it**, which is why no conformance layer reports it today. Worth noting for triage: a guard that exists to dodge a bug is a bug with no test.

## Why this is worth more than its size

The shape is not a corner case. `<div>{props.children}</div>` in a child, and a caller that omits children, is ordinary component code — the first thing someone writes. The failure is user-visible text in the rendered page, and it appears only on the pure-CSR leg, so an SSR-first check would not catch it.

## Owed

Per the repo's three-piece rule this needs a conformance fixture asserting the correct (empty) output with the **unguarded** source, `expectedHtml` generated from Hono (which is correct here), plus a csr-conformance pin carrying this URL until the emission is fixed. Not landed yet — this issue is the first piece.

## Suggested fix

Give the reserved children slot the same undefined handling its siblings already have, at the point the client template is built (`ir-to-client-js`). A `?? ''` at the splice is the minimal form; routing it through the same helper the other props use would be the version that cannot drift apart again — which is the same lesson as #2753/#2762 and #2773.

### PR DESCRIPTION
**Pull request by kfly8**
Fixes #2775.

An unguarded `{props.children}`, called by a parent that passes no children, rendered the literal text `undefined` on a pure-CSR mount while SSR and SSR+hydration rendered nothing. Broken on the SSR ≡ CSR-mount leg only, so an SSR-first check could not see it.

## Emitted template, before and after

Same source both times (`compileJSX` + `TestAdapter`):

```js
// before
template: (_p) => `<div class="card"><h2 bf="s1"><!--bf:s0-->${escapeTextOrMarkup(_p.header)}<!--/--></h2>${_p.children}</div>`
// after
template: (_p) => `<div class="card"><h2 bf="s1"><!--bf:s0-->${escapeTextOrMarkup(_p.header)}<!--/--></h2>${markupOrEmpty(_p.children)}</div>`
```

With no children the key is absent from the props object entirely, so the bare splice stringified `undefined`.

## Why not `escapeTextOrMarkup`

The issue's own suggested fix was to route `children` through the helper its siblings use. That would introduce a worse bug, so this PR does not do it.

`escapeTextOrMarkup` passes a value through unescaped **only** when it carries the `bfMarkup()` brand. A children payload never carries it — `materializeComponent` joins children into a plain HTML string before the template lambda runs ("Template functions expect children as an HTML string, not an array"), and the parent emits it as a raw string:

```js
renderChild('Card', {header: "H", children: `<span>CHILD</span>`}, undefined, 's0')
```

So it would fall through to `escapeText` and render `&lt;span&gt;CHILD&lt;/span&gt;` as visible text on every call that actually has children — trading a bug in the no-children case for one in the common case. `html-template.ts` already records the rule: bare `${...}` interpolations are pre-rendered HTML and must not be escaped.

`markupOrEmpty` therefore does exactly one thing — nullish becomes `''`, everything else passes through untouched. That matches how the other two families already render the same source: Hono's JSX runtime renders `undefined` as nothing, and the DSL adapters route through `bf.string(children)`.

## One door, not four

`bareSpliceExpr` owns the whole decision for the no-`slotId` splice — the counterpart to `escapeTextSlotExpr` for the branch that must *not* escape — and all four `case 'expression'` builders call it. Writing the conditional into each of the four separately would have reproduced the "one decision, N implementations" family this file has already produced twice (#2753 → #2762).

The recognizer is deliberately looser than `isTransparentFragment`, which answers the same question one level up: that function runs on the TS AST and matches an exact set including the analyzer-resolved props name, which is not reachable from the IR layer. A false positive is harmless — the branch never escaped either way, a non-nullish value passes through unchanged, and a nullish one rendering `''` instead of `"undefined"` is an improvement regardless. The comment says so rather than claiming the two are equivalent.

## Pinning instead of routing around

`jsx-element-prop-no-children` and `component-with-jsx-children` both guarded their source with `?? ''`, which is exactly why no conformance layer reported this. Both guards are dropped and `children-passthrough-no-children` is added with the unguarded source, `expectedHtml` generated from Hono unmodified (Hono is correct here). Both docstrings are rewritten so neither rots into a reference to a fixed issue.

## Verification

| Suite | Result |
|---|---|
| `packages/client` | 668 pass / 0 fail |
| `packages/jsx/src` | 3201 pass / 28 skip / 3 todo / 0 fail |
| `packages/adapter-tests/src` | 2241 pass / 4 skip / 1 fail → fixed; `coverage-map.test.ts` 9 pass / 0 fail |
| `packages/compat` | 508 pass / 0 fail (was 507 / 1) |

**Five generated artifacts move with this one emission change.** Each was regenerated with its own tooling and audited against a criterion fixed in advance, rather than accepted because the check went green:

- **Fixture snapshots** — 17 files, 120 changed lines, every one a `markupOrEmpty` insertion; zero added lines without it. `generate-expected-html.ts` then leaves no further drift.
- **`doc-examples.test.ts.snap`** — held 8 stale `${_p.children}` splices, which was the entire cause of the 11 initial jsx failures. Regenerated: 23 insertions, all `markupOrEmpty`, nothing unrelated.
- **`coverage-map.json`** — the two fixtures lose `logical` / `logical:??` / `literal:string` because their guards came out. Checked *before* regenerating that this lowers no floor: 43 other fixtures still cover `logical:??` (including the dedicated `nullish-coalescing-*` and `signal-with-fallback`), the axis count goes 43 → 41, and `uncoveredKinds` is unchanged at `['regex']`.
- **`ui/support-matrix.lock.json`** — audited across all 657 adapter cells: every changed cell moves `pass` and `total` by the same amount in the same direction, and no cell's failure count grows. Totals drop by 2 for the dropped guards and rise by 1 for the new fixture *with `pass` also rising by 1 in every adapter*, so the new fixture passes on all nine, DSL adapters included.
- **`ui/compat.lock.json`** — one line, `totalFixtures` 351 → 352.

The last two are only trustworthy from an environment that can render every adapter for real. Both were verified that way first: all nine adapter packages built, and ruby/go/php/perl/python3 present, with the compat run reporting the same `62 component(s) x 9 adapter(s) = 558 cell(s) (558 ok, 0 with diagnostics)` line as CI. Regenerating `support-matrix:lock` without that gives a silently truncated lock — filed as #2785.

That 17 real components (kbd, dialog, tooltip, select, tabs, popover, combobox, …) were emitting the vulnerable splice is the measurement behind the issue's claim that this is ordinary component code, not a corner case.

### CODE REVIEW

File: packages/jsx/src/ir-to-client-js/html-template.ts [hunk 1]
```diff
@@ -339,6 +349,62 @@ function escapeTextSlotExpr(innerExpr: string, isMarkup = false): string {
...
+ *
+ * That looseness is why the following argument is load-bearing rather than
+ * incidental. A false positive — an unrelated `.children` field, a tree
+ * node's own `children` array say, reaching this bare-splice branch with no
+ * `slotId` — changes nothing harmful: the branch never escaped its value
+ * either way, a non-nullish value is returned untouched, and a nullish one
+ * rendering `''` instead of the literal `"undefined"` is an improvement in
+ * its own right. A destructured-and-renamed children
+ * (`const { children: kids } = props`) is NOT recognized — the same known
+ * gap `isTransparentFragment` has.
```
[the remark this answers is not in the document]
Comment (kfly8) [reply packages/jsx/src/ir-to-client-js/html-template.ts:380]: Correct call, and the convention you cite is the right one — a docstring note is a prose report. Filed as #2786.

Verified the gap first rather than taking the docstring's word for it, and the measurement turned up something the docstring got wrong. Compiled all three shapes and read the emitted `hydrate(... template: ...)` line:

| Source | Emitted splice |
|---|---|
| `props.children` | `${markupOrEmpty(_p.children)}` |
| `{ children }` destructured | `${markupOrEmpty(_p.children)}` |
| `const { children: kids } = props` then `{kids}` | `${(_p.children)}` — unguarded |

The third row is the useful part. The compiler **has already resolved the alias** — the emitted expression is `(_p.children)`, not `kids`. So this is not an inherent limit of the layer the way the docstring implies by analogy to `isTransparentFragment`; the information needed to recognize it is present at the decision point. `bareSpliceExpr` already receives that resolved expression as its second argument and simply does not use it for recognition. Matching on the resolved form should close it, and #2786 records that as the fix direction rather than filing it as a permanent position.

Also measured the current exposure, since "still hits the pre-fix bug" is only alarming if something exercises it: no unguarded `${_p.children}` or `${(_p.children)}` survives in any regenerated fixture snapshot or in `doc-examples.test.ts.snap`, and no component under `ui/components/` or `site/ui/components/` uses a renamed children destructure. Latent gap, not a live one.

Keeping the fix out of this PR deliberately. It is green with five regenerated artifacts each audited against a criterion fixed before running the generator, and the shape is pre-existing rather than introduced here — reopening that verification to carry an unexercised gap would trade real assurance for scope. #2786 carries the repro, the mechanism, and the owed fixture.


---
_Generated by [Claude Code](https://claude.ai/code)_
